options-manager.js

```diff
@@ -0,0 +1,179 @@
+	opts = objectAssign({}, opts);
+	// alias to help humans
+	['env', 'global', 'ignore', 'plugin', 'rule', 'extend'].forEach(function (singular) {
+		var plural = singular + 's';
+		var value = opts[plural] || opts[singular];
+
+		delete opts[singular];
+
+		if (value === undefined) {
+			return;
+		}
+
+		if (singular !== 'rule') {
+			value = arrify(value);
+		}
+
+		opts[plural] = value;
+	});
+
+	return opts;
+}
+
+function mergeWithPkgConf(opts) {
+	opts = objectAssign({
+		cwd: process.cwd()
+	}, opts);
+
+	return objectAssign({}, pkgConf.sync('xo', opts.cwd), opts);
+}
+
+function buildConfig(opts) {
+	var config = deepAssign({}, DEFAULT_CONFIG, {
+		envs: opts.envs,
+		globals: opts.globals,
+		plugins: DEFAULT_PLUGINS.concat(opts.plugins || []),
+		rules: opts.rules,
+		fix: opts.fix
+	});
+
+	if (!config.rules) {
+		config.rules = {};
+	}
+
+	if (opts.space) {
+		var spaces = typeof opts.space === 'number' ? opts.space : 2;
+		config.rules.indent = [2, spaces, {SwitchCase: 1}];
+	}
+
+	if (opts.semicolon === false) {
+		config.rules.semi = [2, 'never'];
+		config.rules['semi-spacing'] = [2, {before: false, after: true}];
+	}
+
+	if (opts.esnext) {
+		config.baseConfig.extends = 'xo/esnext';
+	} else {
+		// always use the Babel parser so it won't throw
+		// on esnext features in normal mode
+		config.parser = 'babel-eslint';
+		config.plugins = ['babel'];
+		config.rules['generator-star-spacing'] = 0;
+		config.rules['arrow-parens'] = 0;
+		config.rules['object-curly-spacing'] = 0;
+		config.rules['babel/object-curly-spacing'] = [2, 'never'];
+	}
+
+	if (opts.extends && opts.extends.length > 0) {
+		// user's configs must be resolved to their absolute paths
+		var configs = opts.extends.map(function (name) {
+			if (name.indexOf('eslint-config-') === -1) {
+				name = 'eslint-config-' + name;
+			}
+
+			return resolveFrom(opts.cwd, name);
+		});
+
+		configs.unshift(config.baseConfig.extends);
+
+		config.baseConfig.extends = configs;
+	}
+
+	return config;
+}
+
+// Builds a list of overrides for a particular path, and a hash value.
+// The hash value is a binary representation of which elements in the `overrides` array apply to the path.
+//
+// If overrides.length === 4, and only the first and third elements apply, then our hash is: 1010 (in binary)
+function findApplicableOverrides(path, overrides) {
+	var hash = 0;
+	var applicable = [];
+	overrides.forEach(function (override) {
+		hash <<= 1;
+		if (mutlimatch(path, override.files).length > 0) {
+			applicable.push(override);
+			hash |= 1;
+		}
+	});
+	return {
+		hash: hash,
+		applicable: applicable
+	};
+}
+
+// Creates grouped sets of merged options together with the paths they apply to.
+function groupConfigs(paths, baseOptions, overrides) {
+	var map = {};
+	var arr = [];
+
+	paths.forEach(function (x) {
+		var data = findApplicableOverrides(x, overrides);
+		if (!map[data.hash]) {
+			var mergedOpts = deepAssign.apply(null, [{}, baseOptions].concat(data.applicable));
```

Comment:
This is where the merging of every applicable "overrides" happens. Is the `deepAssign` logic sufficient here?